design/design-doc-terrajet.md

```diff
@@ -0,0 +1,526 @@
+  * API Output -> Status
+  * Spec <-> API Output Comparison for up-to-date check
+* Controller Code
+  * Readiness Check
+  * Calls to API
+  * Connection Details
+
+| Functionality | Import Provider | Talk to Provider | HCL/JSON |
+| --- | --- | --- | ---|
+| Separate Spec and Status structs | Iterate over schema.Provider | Iterate over schema.Provider | Iterate over schema.Provider |
+| Spec -> API Input | Generated `cty` Translation | Generated `cty` Translation | JSON Marshalling (multi-key) |
+| Target SDK Struct -> Spec Late Initialization | Generated `cty` Translation | Generated `cty` Translation | JSON Marshalling (multi-key) and reflection/code-generation |
+| API Output -> Status | Generated `cty` Translation | Generated `cty` Translation | JSON Marshalling (multi-key) |
+| Spec <-> API Output Comparison for up-to-date check | Generated `cty` Translation (leaking) | Generated `cty` Translation (leaking) | `terraform plan` diff |
+| Readiness Check | `create/update` completion | `create/update` completion | `terraform apply` completion |
+| Calls to API | call `create/update/delete/read` funcs | call `create/update/delete/read` funcs | `terraform apply/destroy` |
+| Connection Details | iterate over `create` result | iterate over `create` result | `sensitive_attributes` in tfstate |
+
+> Note that even if we import provider code, we need to work with `cty` format
+> because the input provider functions expect is of type [`schema.ResourceData`][resource-data],
+> which can be generated only if you have [`terraform.InstanceState`][instance-state]
+> because all of its fields are internal. While state can be constructed, due to
+> its structure of [how it stores][instance-state-map] key/values, using `cty`
+> empowered with the schema of the resource would be the best choice.
+
+> "leaking" in this context means that there are resource-specific details we can't
+> generate.
+
+> jsoniter library allows usage of tag keys other than "json", which would allow
+> us define multiple keys for the same field to cover snake case Terraform struct
+> fields and camel case CRD struct fields.
+
+### Import Provider Go Code
+
+The main advantage of this approach is that we're closer to the actual API calls
+that are made to the external API, which would possibly make things run faster
+since we eliminate HCL middleware.
+
+On the other hand, we'd have to do some operations Terraform CLI does, like
+diff'ing, importing, conversion of JSON/HCL to `cty` and such. Providers do give
+you the customization functions whenever needed, but the main pipeline of execution
+lives in Terraform CLI. So, we would have to reimplement those parts of Terraform
+CLI.
+
+Additionally, the mechanics of providers naturally include core Terraform code
+structs that we'd need to develop logic to work on and that'd increase
+our bug surface since they are not intended to be called by something other than
+Terraform CLI code.
+
```

Comment:
Some providers have begun moving their implementation to `internal/` directories. This may make importing provider go code more complex.

Reply:
Yeah we noticed that with @ulucinar , Azure provider keeps it all under `internal/`. Though I didn't know it was a trend.

We still do some importing to get the schema to build CRD types but that happens in build time so we can put some hacks in place to make it work, like clone-rename to pkg-use.